presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/README.md

```diff
@@ -0,0 +1,162 @@
+actual plan. The simplest possible example of this would be asserting that a
+`SELECT name FROM people` query has an OutputNode with an output symbol that
+actually represents the data in the `name` column of the `people` table.
+
+### Aliases
+
+In order for the plan testing framework to be independent of how Presto
+allocates Symbols, Symbols must be given an alias where they originate, and
+must be referenced by that alias further up the plan.
+
+Binding the alias to the Symbol where it originates must also be independent of
+the SymbolAllocator. This means that the binding must match the right side of
+the assignment; i.e. the value that is being assigned to the symbol.
+
+#### Alias binding example
+
+As an example, take a TableScanNode:
+
+```
+  TableScanNode t {
+    table: TableHandle("people")
+    assignments: { symbol_1: ColumnHandle("name") }
+  }
+```
+
+We can match the value Presto assigns to `symbol_1` by matching a ColumnHandle
+for the `name` column  in the `people` table using `new ColumnReference("people",
+"name")`
+
+Actually creating a binding is done using the special `Matcher` `Alias`.
+
+```
+  Matcher matcher = new Alias("NAME_ALIAS", new ColumnReference("people", "name");
+```
+
+This `Matcher` can then be added to an existing `PlanMatchPattern` using
+`with()`. This would be cumbersome. In practice, any non-trivial node should
+have a corresponding static method in `PlanMatchPattern` to make adding aliases
+more convenient. The complete code to validate the query `SELECT name FROM
+people` using the static methods in PlanMatchPattern would be the following:
+
+```
+output(ImmutableList.of("NAME_ALIAS"),
+	tableScan("people", ImmutableMap.of("NAME_ALIAS", "name"));
+```
+
+### ExpectedValueProviders
+
+In the above example, the Matcher that output() adds to the `PlanMatchPattern`
+for the OutputNode can use the bound alias `NAME_ALIAS` directly to verify that
+the OutputNode contains a symbol that represents the `name` column. For more complex fields in PlanNodes, this isn't generally the case.
+
+As much as possible, fields within PlanNodes should be validated using
+`equals()` methods provided by the objects themselves. This avoids duplicating
+comparison logic in test code, where it's likely to get out of date.
+
+The problem with this is that non-test objects don't understand symbol aliases.
+We can't construct expected values with `Symbols` at the time the tests are
+written, because the names of the `Symbols` are unknown.
+
+Expected values actually have to be created once the aliases of the direct and
+indirect sources of the node you're comparing have been collected by the plan
+matching framework. Doing so requires implementing the
+`ExpectedValueProvider<T>` interface, which the plan matching framework can
+call to get expected values once it can provide the aliases to the
+ExpectedValueProvider.
+
+An implementation of `ExpectedValueProvider<T>` should hold the values and
```

Comment:
I'm not sure this is a best-practice per-se, so much as intrinsic to how they need to work. Is there some other possible way I'm missing?